dev-tools/harness.js

```diff
@@ -0,0 +1,237 @@
+/**
+ * Random Simulation harness for testing and benchmarking purposes.
+ *
+ * Refer to `HARNESS.md` for detailed usage instructions.
+ *
+ * Pokemon Showdown - http://pokemonshowdown.com/
+ *
+ * @license MIT
+ */
+
+'use strict';
+
+const child_process = require('child_process');
+const shell = cmd => child_process.execSync(cmd, {stdio: 'inherit', cwd: __dirname});
+
+// Run the build script if we're being called from the command line.
+// NOTE: `require('../build')` is not safe because `replace` is async.
+if (require.main === module) shell('node ../build');
+
+const BattleStreams = require('../.sim-dist/battle-stream');
+const PRNG = require('../.sim-dist/prng').PRNG;
+const RandomPlayerAI = require('../.sim-dist/examples/random-player-ai').RandomPlayerAI;
+
+// 'move' 70% of the time (ie. 'switch' 30%) and ' mega' 60% of the time that its an option.
+const AI_OPTIONS = {move: 0.7, mega: 0.6};
+
+class Runner {
+	constructor(options) {
+		this.prng = (options.prng && !Array.isArray(options.prng)) ?
+			options.prng : new PRNG(options.prng);
+		this.p1options = Object.assign({}, AI_OPTIONS, options.p1options);
+		this.p2options = Object.assign({}, AI_OPTIONS, options.p2options);
+
+		this.format = options.format;
+		this.input = !!options.input;
+		this.output = !!options.output;
+		this.error = !!options.error;
+	}
+
+	async run() {
+		const battleStream = new RawBattleStream(this.input);
+		const game = this.runGame(this.format, battleStream);
+		if (!this.error) return game;
+		return game.catch(err => {
+			console.log(`\n${battleStream.rawInputLog.join('\n')}\n`);
+			throw err;
+		});
+	}
+
+	async runGame(format, battleStream) {
+		const streams = BattleStreams.getPlayerStreams(battleStream);
+		const spec = {formatid: format, seed: this.prng.seed};
+		const p1spec = this.getPlayerSpec("Bot 1", this.p1options);
+		const p2spec = this.getPlayerSpec("Bot 2", this.p2options);
+
+		const p1 = new RandomPlayerAI(
+			streams.p1, Object.assign({seed: this.newSeed()}, this.p1options)).start();
+		const p2 = new RandomPlayerAI(
+			streams.p2, Object.assign({seed: this.newSeed()}, this.p2options)).start();
+
+		streams.omniscient.write(`>start ${JSON.stringify(spec)}\n` +
+			`>player p1 ${JSON.stringify(p1spec)}\n` +
+			`>player p2 ${JSON.stringify(p2spec)}`);
+
+		let chunk;
+		while ((chunk = await Promise.race([streams.omniscient.read(), p1, p2]))) {
```

Comment:
Why isn't this ``await streams.omniscient.read()`` ?

Reply:
Because then we'd be orphaning the `p1`/`p2` `start` promises.

Reply:
FWIW, The `race` really does appear to be required to catch all errors properly, see https://gist.github.com/scheibo/e044fe2e99b08518ad30b8ab992f24d7 - in both cases we discover the error, but only in the case with `race` can we actually log properly (and continue running afterwards, to potentially find more bugs!). I'm OK with leaving the `race` off to get this submitted, because it will still be useful/identify issues, but having actionable output in the cases where its an AI bug is pretty useful.

Reply:
Oh? Good to know. So far, the bug seems to have been accepted, but without further comments.